lib/sshutils/ctx.go

```diff
@@ -42,24 +46,57 @@ type ConnectionContext struct {
 	// set for all channels.
 	env map[string]string
 
-	// agent is a client to remote SSH agent.
-	agent agent.Agent
+	// forwardAgent indicates that agent forwarding has
+	// been requested for this connection.
+	forwardAgent bool
 
-	// agentCh is SSH channel using SSH agent protocol.
-	agentChannel ssh.Channel
 	// closers is a list of io.Closer that will be called when session closes
 	// this is handy as sometimes client closes session, in this case resources
 	// will be properly closed and deallocated, otherwise they could be kept hanging.
 	closers []io.Closer
+
+	// closed indicates that closers have been run.
+	closed bool
+
+	// cancel cancels the context.Context scope associated with this ConnectionContext.
+	cancel context.CancelFunc
 }
 
 // NewConnectionContext creates a new ConnectionContext instance.
-func NewConnectionContext(nconn net.Conn, sconn *ssh.ServerConn) *ConnectionContext {
+func NewConnectionContext(ctx context.Context, nconn net.Conn, sconn *ssh.ServerConn) *ConnectionContext {
+	ctx, cancel := context.WithCancel(ctx)
 	return &ConnectionContext{
+		Context:    ctx,
 		NetConn:    nconn,
 		ServerConn: sconn,
 		env:        make(map[string]string),
+		cancel:     cancel,
+	}
+}
+
+// StartAgentChannel sets up a new agent forwarding channel against this connection.  The channel
+// is automatically closed when either ConnectionContext, or the supplied context.Context
+// gets canceled.
+func (c *ConnectionContext) StartAgentChannel(ctx context.Context) (agent.Agent, error) {
```

Comment:
All the context plumbing here is added just to clean up the channel.
How about wrapping `agent.Agent` with a new type:
```
type ForwardedAgent struct {
  agent.Agent
  agentChannel io.Closer
}

func (a *ForwardedAgent) Close() error {
  return a.agentChannel.Close()
}
```

Handling the context cancellation from higher levels is still nice, but the caller can deal with it by calling Close in a defer.